src/System.Private.ServiceModel/tests/Common/Infrastructure/src/ConditionalWcfTest.cs

```diff
@@ -10,42 +15,191 @@ namespace Infrastructure.Common
         }
 
         public static bool Domain_Joined()
         {
-            return IsConditionalTestEnabled(TestProperties.Domain_Joined_PropertyName);
+            // If TestProperties does not specify true/false explicitly, we need to autodetect whether
+            // the client and server are domain joined.
+            // Temporarily use the simple heuristic that if we are running the services locally, it is.
+            // Refactor this after integration to address https://github.com/dotnet/wcf/issues/1024 
+            return GetConditionValueFromTestProperties(TestProperties.Domain_Joined_PropertyName,
+                                                       Detect_Server_Is_Local);
         }
 
         public static bool Root_Certificate_Installed()
         {
-            return IsConditionalTestEnabled(TestProperties.Root_Certificate_Installed_PropertyName);
+            // Give precedence to TestProperties, but if the value is unspecified, go ahead and
+            // try to install the root cert now.  Failure auto-detects it cannot be done and skips
+            // subsequent tests.
+            return GetConditionValueFromTestProperties(TestProperties.Root_Certificate_Installed_PropertyName,
+                                                       () =>
+                                                       {
+                                                           try
+                                                           {
+                                                               BridgeClientCertificateManager.InstallRootCertificateFromBridge();
+                                                               return true;
+                                                           }
+                                                           catch
+                                                           {
+                                                               return false;
+                                                           }
+                                                       });
         }
 
         public static bool Client_Certificate_Installed()
         {
-            return IsConditionalTestEnabled(TestProperties.Client_Certificate_Installed_PropertyName);
+            // Give precedence to TestProperties, but if the value is unspecified, go ahead and
+            // try to install the local cert now.  Failure auto-detects it cannot be done and skips
+            // subsequent tests.
+            return GetConditionValueFromTestProperties(TestProperties.Client_Certificate_Installed_PropertyName,
+                                                       () =>
+                                                       {
+                                                           try
+                                                           {
+                                                               BridgeClientCertificateManager.InstallLocalCertificateFromBridge();
```

Comment:
A couple observations worth discussing.
1. Should detection do work? Initially I thought 'no' but then I discovered we had multiple ways of stating a dependency ([ConditionFact(nameof(Root_Certificate_Installed)] and in Endpoints.cs where we do EnsureRootCertificateInstalled on specific endpoints).  So after some thought, I concluded it was fragile to state this dependency using 2 different mechanisms in 2 different places, and [ConditionalFact] became the ideal single place to state a dependency.  Given that, it needed the right to do work that could change the environment.  If we decide this is incorrect, then I want to revisit our use of test helper methods like EnsureRootCertificateInstalled, and we need to have test infrastructure support to allow us to ask "Is the Root Certificate installed?"

Also, bear in mind that we don't have the pre-run setup logic we need to make all the setup happen before tests, so having detection do work is necessary.  I am happy to make detectors only detect once we have infrastructure support to do the work in advance.  This is the initial checkin, like all the rest of the infrastructure changes.

2. Good catch on concurrent dictionary.  I have to say I find CD's design counter-intuitive in this regard. I can convert to a normal dictionary and a lock to prevent this multiple func invoke.